src/number_theory/bertrand.lean

```diff
@@ -0,0 +1,689 @@
+
+lemma factorization_support_subset_filter_prime_range (n : ℕ) :
+  n.factorization.support ⊆ finset.filter nat.prime (finset.range (n + 1)) :=
+begin
+  rw [nat.support_factorization, finset.subset_iff],
+  simp only [list.mem_to_finset, finset.mem_filter, finset.mem_range],
+  intros p p_factor,
+  cases le_or_gt n 0 with n_le_0 n_pos,
+  { rw nat.le_zero_iff at n_le_0,
+    subst n_le_0,
+    simp only [factors_zero, list.not_mem_nil] at p_factor,
+    exfalso, exact p_factor, },
+  { split,
+    { linarith [le_of_dvd n_pos (dvd_of_mem_factors p_factor)], },
+    { exact prime_of_mem_factors p_factor, }, },
+end
+
+lemma prod_subset' {f : ℕ → ℕ} {S T : finset ℕ} (sub : T ⊆ S) (prop : ∀ i ∈ S, ¬(i ∈ T) → f i = 1) :
+  (∏ p in S, f p) = (∏ p in T, f p) :=
+begin
+  rw <-finset.prod_filter_mul_prod_filter_not S (λ i, i ∈ T) f,
+  rw [finset.filter_mem_eq_inter, (finset.inter_eq_right_iff_subset _ _).2 sub],
+  rw finset.prod_filter,
+  simp only [ite_not],
+  have r : ∀ (x ∈ S), ite (x ∈ T) 1 (f x) = 1,
+  { intros x,
+    by_cases x ∈ T,
+    { simp only [ite_eq_left_iff],
+      intros _ bad, exfalso, exact bad h, },
+    { simp,
+      intros x_in_s _,
+      exact prop x x_in_s h, }, },
+  rw finset.prod_congr rfl r,
+  simp,
+end
+
+lemma fooo (n : ℕ) : ∏ p in finset.filter nat.prime (finset.range (n + 1)), p ^ (n.factorization p)
+  = ∏ p in n.factorization.support, p ^ (n.factorization p) :=
+begin
+  refine prod_subset' (factorization_support_subset_filter_prime_range n) _,
+  intros i i_in i_not,
+  simp at i_not,
+  have r : n.factorization i = 0,
+  { rw factorization_eq_zero_iff',
+    by_contradiction,
+    push_neg at h,
+    refine i_not _,
+    exact (mem_factors_iff_dvd h.2.2 h.1).2 h.2.1, },
+  rw [←pow_zero i, r],
+end
+
+-- TODO aesthetically, I think this theorem would be better if the range was 2 * n + 1
+lemma central_binom_factorization (n : ℕ) :
+  ∏ p in finset.filter nat.prime (finset.range (central_binom n + 1)),
+    p ^ ((central_binom n).factorization p)
+  = central_binom n :=
+begin
+  calc ∏ p in finset.filter nat.prime (finset.range (n.central_binom + 1)),
+          p ^ (n.central_binom.factorization p)
+    = ∏ p in (central_binom n).factorization.support, p ^ (n.central_binom.factorization p) :
+      fooo _
+  ... = central_binom n : factorization_prod_pow_eq_self (central_binom_ne_zero _),
+end
+
+lemma sq_prime_is_small' {p n : ℕ} (hp : nat.prime p) (n_big : 2 < n) (small : p ≤ sqrt (2 * n)) :
+  p ^ 2 < 2 * n :=
+begin
+  cases lt_or_ge (p ^ 2) (2 * n),
+  { exact h, },
+  { have small' : p ^ 2 ≤ (2 * n) := le_sqrt'.mp small,
+    have t : p * p = 2 * n := (sq _).symm.trans (small'.antisymm h),
+    have p_even : 2 ∣ p := (or_self _).mp ((prime.dvd_mul prime_two).mp ⟨n, t⟩),
+    rw nat.prime_dvd_prime_iff_eq prime_two hp at p_even,
+    rw [←p_even, sq],
+    exact (mul_lt_mul_left zero_lt_two).mpr n_big, },
+end
+
+lemma sq_prime_is_small {p : ℕ} (hp : nat.prime p) :
```

Comment:
This is only here in c ase for some reason this whole lemma wanted to be extracted somewhere else. It's the iff version of `sq_prime_is_small'`. If that lemma doesn't look suitable for extracting, then this can just be deleted; we only use `sq_prime_is_small'`.

Reply:
I have eliminated this lemma by extracting parts of it and then inlining a proof from those parts.